styx-common/src/main/java/com/spotify/styx/util/ShardedCounter.java

```diff
@@ -0,0 +1,339 @@
+/*-
+ * -\-\-
+ * Spotify Styx Common
+ * --
+ * Copyright (C) 2016 - 2018 Spotify AB
+ * --
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+
+package com.spotify.styx.util;
+
+import com.google.cloud.datastore.Datastore;
+import com.google.cloud.datastore.DatastoreReaderWriter;
+import com.google.cloud.datastore.Entity;
+import com.google.cloud.datastore.EntityQuery;
+import com.google.cloud.datastore.Key;
+import com.google.cloud.datastore.QueryResults;
+import com.google.cloud.datastore.StructuredQuery.PropertyFilter;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.collect.Range;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.ConcurrentModificationException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Random;
+import java.util.stream.IntStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of a resource counter on Datastore. Sharded in a way to support increment,
+ * decrement and capping of a counter to a limit, with increased throughput (compared to the 1 write
+ * per second limit on a single entity group) and strong consistency.
+ *
+ * <p>Note that getCounter is supported specifically _without_ strong consistency.
+ *
+ * <p>Note that the ShardedCounter stores state in the same Datastore database that the caller of this
+ * class can access, too; guarantees don't apply if something else than ShardedCounter updates the
+ * stored state.
+ */
+public class ShardedCounter {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ShardedCounter.class);
+
+  // Ought to be enough (parallelism) for everyone. We could make it dynamic with extra effort.
+  public static final int NUM_SHARDS = 128;
+  public static final Instant CACHE_EXPIRY_DURATION = Instant.ofEpochMilli(1000);
```

Comment:
The suggestion about `Duration` was referring to https://docs.oracle.com/javase/8/docs/api/java/time/Duration.html . `Instant` makes an ill suited type for CACHE_EXPIRY_DURATION